src/System.IO.Pipelines.Networking.Tls/ApplicationProtocols.cs

```diff
@@ -0,0 +1,157 @@
+        {
+            ContextApplicationProtocol protoInfo;
+            Interop.QueryContextAttributesW(ref context, ContextAttribute.ApplicationProtocol, out protoInfo);
+
+            if (protoInfo.ProtoNegoStatus != ApplicationProtocolNegotiationStatus.Success)
+            {
+                throw new InvalidOperationException("Could not negotiate a mutal application protocol");
+            }
+            return GetNegotiatedProtocol(protoInfo.ProtocolId, protoInfo.ProtocolIdSize);
+        }
+
+        public static byte[] GetBufferForProtocolId(ProtocolIds supportedProtocols, bool withAlpnHeader)
+        {
+            short listLength = 0;
+            for (int i = 0; i < _numberOfProtocols.Length; i++)
+            {
+                if (((int) _numberOfProtocols.GetValue(i) & (int) supportedProtocols) > 0)
+                {
+                    listLength += (short) (_allProtocols[i].Length + 1);
+                }
+            }
+            byte[] returnValue;
+            if (withAlpnHeader)
+            {
+                //Now we know the total length of the list lets create our array of protocols and header
+                returnValue = new byte[listLength + _HeaderLength];
+                var spa = new Span<byte>(returnValue);
+                int length = _HeaderLength + listLength - _ContentOffset;
+                spa.Write(length);
+                spa = spa.Slice(4);
+                spa.Write(ApplicaitonProtocolNegotiationExtension.Alpn);
+                spa = spa.Slice(4);
+                spa.Write(listLength);
+                spa = spa.Slice(2);
+                ProtocolList(supportedProtocols, spa);
+            }
+            else
+            {
+                //No header just make an array of the correct size
+                returnValue = new byte[listLength];
+                var spa = new Span<byte>(returnValue);
+                ProtocolList(supportedProtocols, spa);
+            }
+            return returnValue;
+        }
+
+        private static Span<byte> ProtocolList(ProtocolIds supportedProtocols, Span<byte> spa)
+        {
+            for (int i = 0; i < _numberOfProtocols.Length; i++)
+            {
+                if (((int) _numberOfProtocols.GetValue(i) & (int) supportedProtocols) > 0)
+                {
+                    var value = _allProtocols[i];
+                    spa.Write((byte) value.Length);
+                    spa = spa.Slice(1);
+                    var valueSpan = new Span<byte>(value);
+                    valueSpan.CopyTo(spa);
+                    spa = spa.Slice(value.Length);
+                }
+            }
+
+            return spa;
+        }
+
+        internal static unsafe ProtocolIds GetNegotiatedProtocol(byte* protocolId, byte protocolIdSize)
+        {
+            var matchedValue = new Span<byte>(protocolId, protocolIdSize);
+            for (int i = 1; i < _allProtocols.Length; i++)
+            {
+                var proto = _allProtocols[i];
```

Comment:
so... here is my problem I tend to agree, however someone at MS kept saying moar var (naming no names) through code reviews off this repo... and here is the thing the corefx and the asp.net specs clash on this. Corefx says only if the type is obvious... but the Asp.net says any time the compiler lets you.

var flavor = fruit.GetFlavor();

who knows what type flavor is... 

https://github.com/aspnet/Home/wiki/Engineering-guidelines#usage-of-the-var-keyword

So in summary you need to harmonize that between teams

@davidfowl